File: runners/java/src/test/java/dev/openfga/sdk/conformance/steps/ClientSteps.java
package dev.openfga.sdk.conformance.steps;

import dev.openfga.sdk.conformance.support.TestContext;
import dev.openfga.sdk.api.configuration.ClientConfiguration;
import dev.openfga.sdk.api.configuration.Credentials;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import org.assertj.core.api.Assertions;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ClientSteps {
    private final TestContext testContext;

    public ClientSteps(TestContext testContext) {
        this.testContext = testContext;
    }

    @Given("I have a client configured with store {string}")
    public void iHaveAClientConfiguredWithStore(String storeId) throws Exception {
        Map<String, String> config = new HashMap<>();
        config.put("storeId", storeId);
        testContext.createClient(config);
    }

    @Given("I have a client configured with:")
    public void iHaveAClientConfiguredWith(DataTable dataTable) throws Exception {
        Map<String, String> config = new HashMap<>();
        
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            
            switch (key.toLowerCase()) {
                case "apiurl":
                    // ApiUrl is handled in createClient
                    break;
                case "storeid":
                    config.put("storeId", value);
                    break;
                case "authorizationmodelid":
                    config.put("authorizationModelId", value);
                    break;
            }
        }
        
        testContext.createClient(config);
    }

    @Given("I configure authentication with bearer token {string}")
    public void iConfigureAuthenticationWithBearerToken(String token) {
        Map<String, String> config = new HashMap<>();
        config.put("token", token);
        testContext.configureAuthentication("bearer", config);
    }

    @Given("I configure client credentials authentication:")
    public void iConfigureClientCredentialsAuthentication(DataTable dataTable) {
        Map<String, String> credentials = new HashMap<>();
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            credentials.put(key.toLowerCase(), value);
        }
        
        testContext.configureAuthentication("client_credentials", credentials);
    }

    @Given("I configure default headers:")
    public void iConfigureDefaultHeaders(DataTable dataTable) {
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            testContext.getHeaders().put(key, value);
        }
        
        // Apply headers to existing client if available
        if (testContext.getClient() != null) {
            // Note: OpenFGA Java SDK header setting would go here
            // Implementation depends on SDK capabilities
        }
    }

    @Given("I set the request header {string} to {string}")
    public void iSetTheRequestHeader(String key, String value) {
        testContext.setRequestHeader(key, value);
    }

    @Given("I configure retry settings:")
    public void iConfigureRetrySettings(DataTable dataTable) {
        Map<String, Object> retryConfig = new HashMap<>();
        
        List<Map<String, String>> rows = dataTable.asMaps(String.class, String.class);
        for (Map<String, String> row : rows) {
            String key = row.keySet().iterator().next();
            String value = row.get(key);
            
            switch (key.toLowerCase()) {
                case "maxretries":
                    retryConfig.put("maxRetries", Integer.parseInt(value));
                    break;
                case "backofftype":
                    retryConfig.put("backoffType", value);
                    break;
                case "basedelay":
                    retryConfig.put("baseDelay", Integer.parseInt(value));
                    break;
                case "circuitbreakerenabled":
                    retryConfig.put("circuitBreakerEnabled", Boolean.parseBoolean(value));
                    break;
                case "circuitbreakerthreshold":
                    retryConfig.put("circuitBreakerThreshold", Integer.parseInt(value));
                    break;
                case "jitterenabled":
                    retryConfig.put("jitterEnabled", Boolean.parseBoolean(value));
                    break;
            }
        }
        
        // Validate retry configuration
        if (retryConfig.containsKey("maxRetries") && (Integer) retryConfig.get("maxRetries") < 0) {
            throw new IllegalArgumentException("maxRetries must be non-negative");
        }
        
        testContext.getSavedData().put("retry_config", retryConfig);
        testContext.getSavedData().put("retry_attempts", new ArrayList<Long>());
        testContext.getSavedData().put("retry_delays", new ArrayList<Integer>());
    }

    @Then("the client should be configured successfully")
    public void theClientShouldBeConfiguredSuccessfully() {
        Assertions.assertThat(testContext.getClient()).isNotNull();
        Assertions.assertThat(testContext.getConfig()).isNotNull();
    }
}

File: runners/dotnet/StepDefinitions/ClientSteps.cs
using TechTalk.SpecFlow;
using OpenFga.Sdk.Client;
using OpenFga.Sdk.Conformance.Support;
using FluentAssertions;

namespace OpenFga.Sdk.Conformance.StepDefinitions
{
    [Binding]
    public sealed class ClientSteps
    {
        private readonly TestContext _testContext;

        public ClientSteps(TestContext testContext)
        {
            _testContext = testContext;
        }

        [Given(@"I have a client configured with store ""([^""]*)""")]
        public async Task GivenIHaveAClientConfiguredWithStore(string storeId)
        {
            var config = new ClientConfiguration
            {
                StoreId = storeId
            };
            
            await _testContext.CreateClientAsync(config);
        }

        [Given(@"I have a client configured with:")]
        public async Task GivenIHaveAClientConfiguredWith(Table table)
        {
            var config = new ClientConfiguration();
            
            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];
                
                switch (key.ToLowerInvariant())
                {
                    case "apiurl":
                        // ApiUrl is handled in TestContext.CreateClientAsync
                        break;
                    case "storeid":
                        config.StoreId = value;
                        break;
                    case "authorizationmodelid":
                        config.AuthorizationModelId = value;
                        break;
                }
            }
            
            await _testContext.CreateClientAsync(config);
        }

        [Given(@"I configure authentication with bearer token ""([^""]*)""")]
        public async Task GivenIConfigureAuthenticationWithBearerToken(string token)
        {
            var config = new Dictionary<string, string> { ["token"] = token };
            _testContext.ConfigureAuthentication("bearer", config);
        }

        [Given(@"I configure client credentials authentication:")]
        public async Task GivenIConfigureClientCredentialsAuthentication(Table table)
        {
            var credentials = new Dictionary<string, string>();
        
            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];
                credentials[key.ToLowerInvariant()] = value;
            }
        
            _testContext.ConfigureAuthentication("client_credentials", credentials);
        }

        [Given(@"I configure default headers:")]
        public async Task GivenIConfigureDefaultHeaders(Table table)
        {
            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];
                _testContext.Headers[key] = value;
            }
            
            // Apply headers to existing client if available
            if (_testContext.Client != null)
            {
                // Note: OpenFGA .NET SDK header setting would go here
                // Implementation depends on SDK capabilities
            }
        }

        [Given(@"I set the request header ""([^""]*)"" to ""([^""]*)""")]
        public void GivenISetTheRequestHeader(string key, string value)
        {
            _testContext.SetRequestHeader(key, value);
        }

        [Given(@"I configure retry settings:")]
        public void GivenIConfigureRetrySettings(Table table)
        {
            var retryConfig = new Dictionary<string, object>();
            
            foreach (var row in table.Rows)
            {
                var key = row.Keys.First();
                var value = row[key];
                
                switch (key.ToLowerInvariant())
                {
                    case "maxretries":
                        retryConfig["MaxRetries"] = int.Parse(value);
                        break;
                    case "backofftype":
                        retryConfig["BackoffType"] = value;
                        break;
                    case "basedelay":
                        retryConfig["BaseDelay"] = int.Parse(value);
                        break;
                    case "circuitbreakerenabled":
                        retryConfig["CircuitBreakerEnabled"] = bool.Parse(value);
                        break;
                    case "circuitbreakerthreshold":
                        retryConfig["CircuitBreakerThreshold"] = int.Parse(value);
                        break;
                    case "jitterenabled":
                        retryConfig["JitterEnabled"] = bool.Parse(value);
                        break;
                }
            }
            
            // Validate retry configuration
            if (retryConfig.ContainsKey("MaxRetries") && (int)retryConfig["MaxRetries"] < 0)
            {
                throw new ArgumentException("maxRetries must be non-negative");
            }
            
            _testContext.SavedData["retry_config"] = retryConfig;
            _testContext.SavedData["retry_attempts"] = new List<DateTime>();
            _testContext.SavedData["retry_delays"] = new List<int>();
        }

        [Then(@"the client should be configured successfully")]
        public void ThenTheClientShouldBeConfiguredSuccessfully()
        {
            _testContext.Client.Should().NotBeNull();
            _testContext.Config.Should().NotBeNull();
        }
    }
}
